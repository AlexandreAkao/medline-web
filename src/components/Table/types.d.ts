interface IColumn {
  Header: string;
  accessor: string;
  columns?: IColumn[];
  Cell?: (props: import('react-table').UseTableCellProps<unknown>) => React.ReactNode;
}

interface IData {
  [x: string]: React.ReactNode;
}

interface ITableProps {
  columns: IColumn[];
  data: IData[];
}
