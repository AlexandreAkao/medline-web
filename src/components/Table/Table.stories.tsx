import { ComponentStory, ComponentMeta } from '@storybook/react';
import { AiOutlineEdit } from 'react-icons/ai';

import Table from 'components/Table';

export default {
  title: 'Table',
  component: Table,
  decorators: [
    Story => (
      <div
        style={{
          padding: 20,
        }}
      >
        <Story />
      </div>
    ),
  ],
} as ComponentMeta<typeof Table>;

const Template: ComponentStory<typeof Table> = args => <Table {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  columns: [
    {
      Header: 'First Name',
      accessor: 'firstName',
    },
    {
      Header: 'Last Name',
      accessor: 'lastName',
    },
    {
      Header: 'Age',
      accessor: 'age',
    },
    {
      Header: '',
      accessor: 'action',
      Cell: ({ value }) => {
        if (typeof value === 'function') return value();

        return value;
      },
    },
  ],
  data: [
    {
      firstName: 'FirstName1',
      lastName: 'LastName1',
      age: 8,
      action: () => <AiOutlineEdit />,
    },
    {
      firstName: 'FirstName2',
      lastName: 'LastName2',
      age: 20,
      action: () => <AiOutlineEdit />,
    },
    {
      firstName: 'FirstName3',
      lastName: 'LastName3',
      age: 100,
      action: () => <AiOutlineEdit />,
    },
  ],
};
