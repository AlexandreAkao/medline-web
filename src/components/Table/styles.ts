import styled from 'styled-components';

import colors from 'styles/colors';

export const TableContainer = styled.div`
  table {
    border-spacing: 0;
    box-shadow: 0px 0px 10px 0px rgba(0, 0, 0, 0.2);
    border-radius: 12px;
    width: 100%;
    overflow: hidden;
    border-collapse: separate;

    thead {
      tr {
        background-color: ${colors.primary.lighter};
      }
    }

    tbody {
      tr {
        :nth-child(even) {
          background-color: ${colors.white.lighter};
        }
        :nth-child(odd) {
          background-color: ${colors.white.normal};
        }

        :last-child {
          td {
            border-bottom: 0;
          }
        }
      }
    }

    th,
    td {
      margin: 0;
      padding: 0.5rem;
      border-bottom: 2px solid ${colors.white.lightest};
      border-right: 2px solid ${colors.white.lightest};

      :last-child {
        border-right: 0;
        text-align: center;
      }
    }

    th {
      padding: 1rem;
    }
  }
`;
