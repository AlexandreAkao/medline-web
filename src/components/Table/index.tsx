import { useTable } from 'react-table';

import { TableContainer } from 'components/Table/styles';

function Table({ columns, data }: ITableProps) {
  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } = useTable({
    columns,
    data,
  });

  return (
    <TableContainer>
      <table {...getTableProps()}>
        <thead>
          {headerGroups.map(headerGroup => (
            <tr {...headerGroup.getHeaderGroupProps()} key={headerGroup.getHeaderGroupProps().key}>
              {headerGroup.headers.map(column => (
                <th {...column.getHeaderProps()} key={column.getHeaderProps().key}>
                  {column.render('Header')}
                </th>
              ))}
            </tr>
          ))}
        </thead>
        <tbody {...getTableBodyProps()}>
          {rows.map(row => {
            prepareRow(row);
            return (
              <tr {...row.getRowProps()} key={row.id}>
                {row.cells.map(cell => {
                  return (
                    <td {...cell.getCellProps()} key={cell.getCellProps().key}>
                      {cell.render('Cell')}
                    </td>
                  );
                })}
              </tr>
            );
          })}
        </tbody>
      </table>
    </TableContainer>
  );
}

export default Table;
