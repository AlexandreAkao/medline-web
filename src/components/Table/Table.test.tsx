import { render, screen } from '@testing-library/react';
import { composeStories } from '@storybook/testing-react';

import * as stories from 'components/Table/Table.stories';
import Table from 'components/Table';

const { Primary } = composeStories(stories);

describe('Table', () => {
  const renderComponent = (customProps: Partial<ITableProps> = {}) => {
    const props: ITableProps = {
      columns: [],
      data: [],
      ...customProps,
    };
    return render(<Table {...props} />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly', () => {
    render(<Primary />);
    const Header1Element = screen.getByText(/First Name/i);
    const Header2Element = screen.getByText(/Last Name/i);
    const Header3Element = screen.getByText(/Age/i);
    const Cell1Element = screen.getByText(/FirstName1/i);
    const Cell2Element = screen.getByText(/FirstName2/i);
    const Cell3Element = screen.getByText(/FirstName3/i);

    expect(Header1Element).toBeInTheDocument();
    expect(Header2Element).toBeInTheDocument();
    expect(Header3Element).toBeInTheDocument();
    expect(Cell1Element).toBeInTheDocument();
    expect(Cell2Element).toBeInTheDocument();
    expect(Cell3Element).toBeInTheDocument();
  });

  it('should render a table with correctly rows', () => {
    renderComponent({
      columns: [
        { Header: 'Header1', accessor: 'Header1' },
        { Header: 'Header2', accessor: 'Header2' },
      ],
      data: [{ Header1: 'Cell1', Header2: 'Cell2' }],
    });

    const rows = screen.getAllByRole('row');

    expect(rows).toHaveLength(2);
  });
});
