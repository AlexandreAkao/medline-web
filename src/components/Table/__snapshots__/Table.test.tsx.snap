// Vitest Snapshot v1

exports[`Table > should match with snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ duHOFq"
  >
    <table
      role="table"
    >
      <thead />
      <tbody
        role="rowgroup"
      />
    </table>
  </div>
</div>
`;
